ssl/s3_lib.c

```diff
@@ -3574,16 +3579,82 @@ long ssl3_ctrl(SSL *s, int cmd, long larg, void *parg)
         break;
 
     case SSL_CTRL_GET_TLSEXT_STATUS_REQ_OCSP_RESP:
-        *(unsigned char **)parg = sc->ext.ocsp.resp;
-        if (sc->ext.ocsp.resp_len == 0
-                || sc->ext.ocsp.resp_len > LONG_MAX)
-            return -1;
-        return (long)sc->ext.ocsp.resp_len;
+        *(unsigned char **)parg = NULL;
+        ret = -1;
+
+#ifndef OPENSSL_NO_OCSP
+        if (sc->ext.ocsp.resp != NULL) {
+            OPENSSL_free(sc->ext.ocsp.resp);
+            sc->ext.ocsp.resp = NULL;
+            sc->ext.ocsp.resp_len = 0;
+        }
+
+        resp = sk_OCSP_RESPONSE_value(sc->ext.ocsp.resp_ex, 0);
+
+        if (resp != NULL) {
+            int resp_len = i2d_OCSP_RESPONSE(resp, NULL);
+
+            if (resp_len > 0 && resp_len <= LONG_MAX) {
+                sc->ext.ocsp.resp = OPENSSL_malloc(resp_len);
+
+                if (sc->ext.ocsp.resp != NULL) {
+                    *(unsigned char **)parg = sc->ext.ocsp.resp;
+                    p = sc->ext.ocsp.resp;
+                    sc->ext.ocsp.resp_len = i2d_OCSP_RESPONSE(resp, &p);
+                    ret = sc->ext.ocsp.resp_len;
+                } else {
+                    ret = -1;
+                }
+            }
+        } else {
+            ret = -1;
+        }
+#endif
+        break;
 
     case SSL_CTRL_SET_TLSEXT_STATUS_REQ_OCSP_RESP:
-        OPENSSL_free(sc->ext.ocsp.resp);
-        sc->ext.ocsp.resp = parg;
-        sc->ext.ocsp.resp_len = larg;
+#ifndef OPENSSL_NO_OCSP
+        if (sc->ext.ocsp.resp != NULL) {
+            OPENSSL_free(sc->ext.ocsp.resp);
+            sc->ext.ocsp.resp = NULL;
+            sc->ext.ocsp.resp_len = 0;
+        }
+        sk_OCSP_RESPONSE_pop_free(sc->ext.ocsp.resp_ex, OCSP_RESPONSE_free);
+        sc->ext.ocsp.resp_ex = NULL;
+
+        if (parg != NULL) {
+            sc->ext.ocsp.resp_ex = sk_OCSP_RESPONSE_new_null();
+
+            p = parg;
+            resp = d2i_OCSP_RESPONSE(NULL, (const unsigned char **)&p, larg);
+            sk_OCSP_RESPONSE_push(sc->ext.ocsp.resp_ex, resp);
+        }
+#endif
+        ret = 1;
```

Comment:
You should free `parg`.

Reply:
No. This needs to be fixed.  In case of success, the original API takes over the parg so that it is no longer owned by the caller. The caller must never free it in case of success.

If you are afraid of freeing it too early, why don't you just assign it to `sc->ext.ocsp.resp`?